distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,781 @@
+        self._size = 0
+
+    def _buffer_append(self, data):
+        data = memoryview(data).cast("B")
+        self._buffers.append(data)
+        self._size += len(data)
+
+    def _buffer_peek(self):
+        return self._buffers[0][self._first_pos :]
+
+    def _buffer_advance(self, size):
+        b = self._buffers[0]
+        self._size -= size
+        b_len = len(b) - self._first_pos
+        if b_len == size:
+            self._buffers.popleft()
+            self._first_pos = 0
+        else:
+            self._first_pos += size
+
+    def write(self, data):
+        transport = self.transport
+
+        if transport._eof:
+            raise RuntimeError("Cannot call write() after write_eof()")
+        if transport._empty_waiter is not None:
+            raise RuntimeError("unable to write; sendfile is in progress")
+        if not data:
+            return
+        if transport._conn_lost:
+            return
+
+        if not self._buffers:
+            try:
+                n = transport._sock.send(data)
+            except (BlockingIOError, InterruptedError):
+                pass
+            except (SystemExit, KeyboardInterrupt):
+                raise
+            except BaseException as exc:
+                transport._fatal_error(exc, "Fatal write error on socket transport")
+                return
+            else:
+                data = data[n:]
+                if not data:
+                    return
+            # Not all was written; register write handler.
+            transport._loop._add_writer(transport._sock_fd, self._on_write_ready)
+
+        # Add it to the buffer.
+        self._buffer_append(data)
+        transport._maybe_pause_protocol()
+
+    def writelines(self, buffers):
+        waiting = bool(self._buffers)
+        for b in buffers:
+            self._buffer_append(b)
+        if not waiting:
+            try:
+                self._do_bulk_write()
+            except (BlockingIOError, InterruptedError):
+                pass
+            except (SystemExit, KeyboardInterrupt):
+                raise
+            except BaseException as exc:
+                self.transport._fatal_error(
+                    exc, "Fatal write error on socket transport"
+                )
+                return
+            if not self._buffers:
+                return
+            # Not all was written; register write handler.
+            self.transport._loop._add_writer(
+                self.transport._sock_fd, self._on_write_ready
+            )
+
+        self.transport._maybe_pause_protocol()
+
+    def is_closing(self):
+        return self.transport.is_closing()
+
+    def close(self):
+        return self.transport.close()
+
+    def abort(self):
+        return self.transport.abort()
+
+    def get_extra_info(self, key):
+        return self.transport.get_extra_info(key)
+
+    def _do_bulk_write(self):
+        buf = self._buffer_peek()
+        n = self.transport._sock.send(buf)
```

Comment:
A previous commit made use of `sendmsg` to write multiple buffers at once. This was actually _slower_ in most cases, which I found surprising. Fewer syscalls were made, but each syscall took longer :shrug:.